sw/device/lib/base/status.h

```diff
@@ -49,42 +52,66 @@ extern "C" {
-#define TRY(expr_)            \
-  ({                          \
-    status_t status_ = expr_; \
-    if (status_.value < 0) {  \
-      return status_;         \
-    }                         \
-    status_.value;            \
+/**
+ * Converts a value into a status_t.
+ *
+ * This macro uses the GCC/clang `__builtin_types_compatible_p` extension to
+ * detect the type of the input expression and apply the appropriate
+ * conversion.  Once a more thorough refactoring of the DIFs is done, this
+ * can be eliminated.
+ *
+ * @param expr_ Either a `status_t`, `rom_error_t` or `dif_result_t`.
+ * @return The `status_t` representation of the input.
+ */
+#define INTO_STATUS(expr_)                                                \
+  ({                                                                      \
+    typeof(expr_) ex_ = (expr_);                                          \
+    status_t status_;                                                     \
+    if (__builtin_types_compatible_p(typeof(ex_), status_t)) {            \
+      memcpy(&status_, &ex_, sizeof(status_));                            \
+    } else if (__builtin_types_compatible_p(typeof(ex_), rom_error_t)) {  \
+      memcpy(&status_, &ex_, sizeof(status_));                            \
+    } else if (__builtin_types_compatible_p(typeof(ex_), dif_result_t)) { \
+      absl_status_t code;                                                 \
+      memcpy(&code, &ex_, sizeof(code));                                  \
+      status_ = status_create(code, status_module_id, __FILE__,           \
+                              code == kOk ? 0 : __LINE__);                \
+    } else {                                                              \
+      static_assert(                                                      \
+          __builtin_types_compatible_p(typeof(ex_), status_t) ||          \
+              __builtin_types_compatible_p(typeof(ex_), rom_error_t) ||   \
+              __builtin_types_compatible_p(typeof(ex_), dif_result_t),    \
+          "Expressions passed to INTO_STATUS() must be of type "          \
+          "`status_t`, `rom_error_t` or `dif_result_t`");                 \
+    }                                                                     \
+    status_;                                                              \
+  })
+
+/**
+ * Evaluates a status_t for Ok or Error status, returning the Ok value.
+ *
+ * This macro is like the `try!` macro (or now `?` operator) in Rust:
+ * It evaluates to the contained OK value or it immediately returns from
+ * the enclosing function with the error value.
+ *
+ * @param expr_ An expression that can be converted to a `status_t`.
+ * @return The enclosed OK value.
+ */
+#define TRY(expr_)                         \
+  ({                                       \
+    status_t status_ = INTO_STATUS(expr_); \
+    if (status_.value < 0) {               \
```

Comment:
nit: `status_.value & STATUS_BIT_ERROR`?

Reply:
nit on your nit: `status_.value & (1 << STATUS_BIT_ERROR)`.

The comparison with zero ensures that no constant (like 0x80000000) needs to be loaded or used as an operand in the generated assembly (on either rv32i or x86).

Reply:
> nit on your nit: `status_.value & (1 << STATUS_BIT_ERROR)`.

Hehe, sorry about that :)

> The comparison with zero ensures that no constant (like 0x80000000) needs to be loaded or used as an operand in the generated assembly (on either rv32i or x86).

SG, I didn't realize that the `value` field was an `int32_t`.